File: java/planner/src/main/java/com/risingwave/scheduler/task/RemoteTaskManager.java
package com.risingwave.scheduler.task;

import static java.util.Objects.requireNonNull;

import akka.actor.typed.ActorRef;
import akka.actor.typed.Behavior;
import akka.actor.typed.javadsl.Behaviors;
import com.risingwave.rpc.ComputeClientManager;
import com.risingwave.scheduler.EventListener;
import com.risingwave.scheduler.actor.ActorFactory;
import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
public class RemoteTaskManager implements TaskManager {
  private final ComputeClientManager clientManager;

  private final ActorRef<TaskManagerEvent> actor;

  @Inject
  public RemoteTaskManager(ComputeClientManager clientManager, ActorFactory actorCreator) {
    this.clientManager = requireNonNull(clientManager, "clientManager");

    Behavior<TaskManagerEvent> taskManagerActor =
        Behaviors.setup(ctx -> new TaskManagerActor(ctx, clientManager));
    this.actor = actorCreator.createActor(taskManagerActor, "RemoteTaskManager");
  }

  @Override
  public void schedule(QueryTask task, EventListener<TaskEvent> taskEventListener) {
    actor.tell(new TaskManagerEvent.ScheduleTaskEvent(task, taskEventListener));
  }
}

File: java/planner/src/main/kotlin/com/risingwave/scheduler/task/RemoteTaskManager.kt
package com.risingwave.scheduler.task

import com.risingwave.node.WorkerNode
import com.risingwave.proto.common.Status
import com.risingwave.proto.computenode.CreateTaskRequest
import com.risingwave.proto.computenode.CreateTaskResponse
import com.risingwave.rpc.ComputeClientManager
import org.slf4j.LoggerFactory
import java.util.Objects
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class RemoteTaskManager @Inject constructor(
  clientManager: ComputeClientManager
) : TaskManager {

  companion object {
    private val log = LoggerFactory.getLogger(RemoteTaskManager::class.java)
  }

  private val clientManager: ComputeClientManager

  init {
    this.clientManager = Objects.requireNonNull(clientManager, "clientManager")
  }

  override suspend fun schedule(task: QueryTask): WorkerNode {
    val taskId: TaskId = task.taskId
    val node: WorkerNode = task.queryStage.workers[taskId.id]
    val response = sendCreateTaskRequest(task, node)
    if (response.status.code != Status.Code.OK) {
      log.error("Failed to create task: ${response.status.code}, ${response.status.message}")
    } else {
      log.info("Succeeded to create task: ${task.taskId}")
    }
    return node
  }

  private suspend fun sendCreateTaskRequest(task: QueryTask, node: WorkerNode): CreateTaskResponse {
    val client = clientManager.getOrCreate(node)
    val request: CreateTaskRequest = CreateTaskRequest.newBuilder()
      .setTaskId(task.taskId.toTaskIdProto())
      .setPlan(task.queryStage.toPlanFragmentProto())
      .build()
    return client.createTaskKt(request)
  }
}
